syntax = "proto3";

import "farm_ng/perception/apriltag.proto";
import "farm_ng/perception/camera_model.proto";
import "farm_ng/perception/geometry.proto";
import "farm_ng/perception/image.proto";
import "farm_ng/calibration/calibrator.proto";

package farm_ng.calibration;
option go_package = "github.com/farm-ng/genproto/calibration";



message IntrinsicModel {
  // The solved for camera model.
  farm_ng.perception.CameraModel camera_model = 1;

  // Apriltag rigs used for solving camera model
  repeated farm_ng.perception.ApriltagRig apriltag_rigs = 2;

  // Is this the initialization or did the solver converge, fail...
  // For diagnosis of failures or unexpected results its useful to look at the
  // initialization first.
  SolverStatus solver_status = 3;
  // The total rig RMSE of all tags from all images.
  double rmse = 4;

  // Per tag statistics
  repeated ApriltagRigTagStats tag_stats = 5;

  // The ApriltagDetections that are used to initialize and solve the
  // camera model. For debugging purposes, these typically have the image resources
  // associated with them.  The tag_stats.per_image_rmse keys refer to the
  // elements of this collection.
  repeated farm_ng.perception.ApriltagDetections detections = 6;

  // The solver also estimates the pose of each of the camera views w.r.t to the apriltag rigs.
  //  This is may not be the same length of
  // detections, for example if the detections had outliers or not enough
  // detections to estimate the pose, but the frame names
  // <camera_frame_name>/<detections_index %05d> encode
  // encode which index in detections these refer to.
  repeated farm_ng.perception.NamedSE3Pose camera_poses_rig = 7;

  // Each image where we are able to estimate a pose has a projection debug
  // image, which shows the detected points, and the reprojected corners of the
  // rig.
  repeated farm_ng.perception.Image reprojection_images = 8;
}
